### PR TITLE
feat: queryClient.ensureQueryData

### PR DESCRIPTION
**Pull request by zorzysty**
Shorthand for calling `queryClient.getQueryData(...) ?? queryClient.fetchQuery(...)`

### CODE REVIEW

File: packages/query-core/src/queryClient.ts [hunk 1]
```diff
@@ -118,6 +118,21 @@ export class QueryClient {
...
   }
 
+  ensureQueryData<
+    TQueryFnData = unknown,
+    TQueryKey extends QueryKey = QueryKey,
+  >(
+    queryKey: TQueryKey,
+    queryFn: QueryFunction<TQueryFnData, TQueryKey>,
+    filters?: QueryFilters,
+  ): Promise<TQueryFnData> {
```
Comment (TkDodo) [inline packages/query-core/src/queryClient.ts:128]: I think we need the same interface here as `fetchQuery`. That means:

- 3 overloads (we'll remove them in v5)
  - if you want to do me a favour, you'll write the tests with the object-only syntax :)
- the same interface as `fetchQuery`: 4 generics and `FetchQueryOptions`
- we probably don't need the `filters` on `getQueryData`. they are only needed so that you can opt-out of exact filtering and add some other fuzzy filter options instead. It doesn't really fit the use-case for `ensureQueryData`.

Comment (zorzysty) [reply packages/query-core/src/queryClient.ts:128]: Oh, I thought it was suppose to have a signature based on `getQueryData`. I'll change it, no problem.

File: docs/reference/QueryClient.md [hunk 2]
```diff
@@ -178,6 +179,22 @@ const data = queryClient.getQueryData({ queryKey })
...
+
+`ensureQueryData` is an asynchronous function that can be used to get an existing query's cached data. If the query does not exist, `queryClient.fetchQuery` will be called and its results returned.
+
+```tsx
+const data = queryClient.ensureQueryData({ queryKey, queryFn })
+```
+
+**Options**
+
+- The options for `ensureQueryData` are exactly the same as those of [`fetchQuery`](#queryclientfetchquery).
```
Comment (TkDodo) [inline docs/reference/QueryClient.md:192]: except that `queryKey` is required maybe?

File: packages/query-core/src/queryClient.ts [hunk 3]
```diff
@@ -118,6 +119,61 @@ export class QueryClient {
...
+  ): Promise<TData>
+  ensureQueryData<
+    TQueryFnData = unknown,
+    TError = unknown,
+    TData = TQueryFnData,
+    TQueryKey extends QueryKey = QueryKey,
+  >(
+    queryKey: TQueryKey,
+    queryFn: QueryFunction<TQueryFnData, TQueryKey>,
+    options?: FetchQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
```
Comment (TkDodo) [inline packages/query-core/src/queryClient.ts:150]: in the overloads with multiple parameters, we could `omit` the extra parameters from the `options` because it doesn't make sense to pass `queryKey` or `queryFn` in `options` if we have them as a separate parameter.

I did that for `useQuery`, but for nothing else. Maybe it could be a separate PR. Or maybe it doesn't matter because it will be simplified with v5 anyways 😅

File: docs/reference/QueryClient.md (hunk 2 above)
Comment (zorzysty) [reply docs/reference/QueryClient.md]: [b0537ed](https://github.com/TanStack/query/pull/4544/commits/b0537ed35af9dcdbb08f0765cb3bbc67499048f4)

File: packages/query-core/src/queryClient.ts (hunk 3 above)
Comment (zorzysty) [reply packages/query-core/src/queryClient.ts]: [5abe28e](https://github.com/TanStack/query/pull/4544/commits/5abe28e4c78fcf5a66aaa63dc73a6783f48cb809)
I don't think it can be changed for `arg2` and `arg3` without touching the implementation of `parseQueryArgs`. But for the second and third overload it seems to be working fine.

File: packages/query-core/src/queryClient.ts [hunk 4]
```diff
@@ -118,6 +119,67 @@ export class QueryClient {
...
+    TQueryFnData = unknown,
+    TError = unknown,
+    TData = TQueryFnData,
+    TQueryKey extends QueryKey = QueryKey,
+  >(
+    queryKey: TQueryKey,
+    queryFn: QueryFunction<TQueryFnData, TQueryKey>,
+    options?: Omit<
+      FetchQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
+      'queryKey'
```
Comment (TkDodo) [inline packages/query-core/src/queryClient.ts:155]:
```suggestion
      'queryKey' | 'queryFn'
```

**Review by tannerlinsley — APPROVED** [review]

**Comment by XpamAmAdEuS** [comment]
It would be great if we also have same shorthand for prefetchQuery.

**Comment by TkDodo** [comment]
> It would be great if we also have same shorthand for prefetchQuery.

add `.catch(noop)` to it and you got it 🤷‍♂️
